### ISSUE TITLE
Arbitrary operation masking

### ISSUE THREAD

**Issue by Ben-Lichtman** [rust-lang/portable-simd#202]
Hi, I was wondering if this team is considering the addition of masking arbitrary operations rather than just the gather select?

This is something already available in AVX512 through the kmask registers, though would probably require emulation on other platforms, but it seems like a very useful thing to have if at all possible.

One of the more awesome use cases I've seen for this is with vectorised emulation: https://gamozolabs.github.io/fuzzing/2018/10/14/vectorized_emulation.html which makes heavy use of masking to run / fuzz 8 virtual machines in parallel per core per thread.

**Comment by calebzulawski** [comment]
The recommended way to do this is to simply perform the operation and mask as desired--when producing the asm, the backend will recognize that the two operations can be performed with a single instruction.

**Comment by Ben-Lichtman** [comment]
Ah, I see. That's reasonable then

**Comment by programmerjake** [comment]
this could still be reopened, cuz many operations still need masked versions, such as masked integer division, masked shift, etc.

**Comment by workingjubilee** [comment]
re: shifts, I believe we should simply default our SIMD shift to the "wrapping" semantic the same way we default other operations to wrapping, also see https://github.com/rust-lang/rust/issues/91237

I would be surprised if `shl_or` yielded better perf than `wrapping_simd_shl`.

**Comment by workingjubilee** [comment]
This is a duplicate of https://github.com/rust-lang/portable-simd/issues/118,
but I am arbitrarily closing that one instead because this has more actual content.

**Comment by workingjubilee** [comment]
As of the sync getting merged, the shift ops on `Simd` types will wrap, like most of our ops do.
